Alternate setup-version implementation

wrouesnel opened the pull request:
Splits setup-version into the worker script and a library bash script to allow easier code reuse and adaptation between different CI systems.

Integrates a basic test system to ensure correct functioning under CI conditions (population of environment variables).

stephanh commented on lib-version:
```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
```
I was considering a while ago that we should rewrite most of our scripts as functions instead of stand alone shell scripts. With that in mind, what do you think of changing this file to be be called `lib-ci` so it's more generic.

wrouesnel replied:
Do we want to have all the functions in one include file though? `lib-version` == version strings, maybe split it to `lib-scala` and `lib-haskell` ?

stephanh replied:
For builds, do we want to run separate shell scripts or just functions?

If functions than we should aim to only have to source 1 file in any given build (so we could have different files for scala, haskell builds).

Given the initial version string, how to derive the new version should be the same across all builds so we should have a shared function for that used by the other scripts.

wrouesnel replied:
In favor of `setup-version.bsh` is it provides the minimal amount of CI needed (versioning) without needing to commit a whole new script? Note that internally it just calls `get_version` -> maybe move the error checking / file stuff inside that function ?

stephanh commented on lib-version:
```diff
@@ -0,0 +1,85 @@
+
+# Determine the current CI environment
+function get_CI_env() {
+    if [ "$CI" = "true" ]; then
```
Not sure this works. I think both travis and drone set `CI`.

stephanh replied:
I would also say that as part of the generic ci checking we should check that all the CI vars exist, otherwise we can't support the generic ci.

wrouesnel replied:
I'm going to consider this addressed since all the vars are now explicit.

stephanh commented on lib-version:
```diff
@@ -0,0 +1,85 @@
+    "BUILD_DIR" \
+    "BUILD_URL" \
+    "TAG" \
+)
```
Can we remove all the vars that we don't currently use, that way compatibility is much easier.

wrouesnel replied:
Forward tracking to figure out what we do and don't use is kind of unmaintainable. I'd rather just determine what we think we *should* have and make sure it's there. This set seems pretty reasonable.

wrouesnel commented:
I'm going to set a travis build which runs the tests - that'll determine if the variables get found.

stephanh commented on lib-version:
```diff
@@ -0,0 +1,85 @@
+}
+
+# Export CI environment to the common format
+function adapt_CI_env() {
```
Can you maybe expand on this comment to say that all the CI_VARS should now be exported as `CI_NAME`, etc.

wrouesnel replied:
These have been removed and are replaced with explicit variable names now. So we only depend on `DRONE_` and `TRAVIS_` prefixes, more can be added easily.

stephanh commented on lib-version:
```diff
@@ -0,0 +1,85 @@
+    done
+}
+
+# Get a version number based on the Omnia standard
```
Please define what the omnia standard is here.

* `major.minor.path-commish-timestamp` for master branch
* `major.minor.path-commish-timestamp-PRNN` for pull request NN
*  `major.minor.path-commish-timestamp-branchname` for branch `branchname`.

stephanh commented:
@jsok can you have a look at this as well.

stephanh commented on t/test-lib-version.bsh:
```diff
@@ -0,0 +1,38 @@
+# get_version tests
+export CI_COMMIT=adc83b19e793491b1c6ea0fd8b46cd9f32e592fc
+export CI_BRANCH=master
+export CI_PULL_REQUEST=false
```
Can we just locally set them instead of exporting them so that they won't interfere with any CI we might add for this repo?

wrouesnel replied:
Those exports are inside the standalone test script. They shouldn't percolate above the script - we have to export in order to make them available to the subshells.

stephanh commented:
@wrouesnel can you please add a .travis.yml file for this as well since it is hosted on Github and run by travis.

wrouesnel commented:
Travis build confirms the issue. I'll remove the "CI_" class of variables I think.

wrouesnel commented:
Squashed and rebased into a single commit. Travis is less cooperative then hoped, so I decided to just make all the mappings explicit.

wrouesnel commented:
Closed and reopened to trigger CI.

wrouesnel commented:
Okay so here's what I've done:

`lib-version` -> `lib-ci`, library files go under the `lib` directory.

Renamed the functions to follow a more predictable naming scheme, so we get:
 * CI_Env_Get
 * CI_Env_Dump
 * CI_Env_Adapt
 * Version_Get

They're slightly unwieldy to type, but they stand out and clearly show how they relate to things I think. `setup-version.bsh` is converted to use the new names.

I've also moved the version format checking inside the Version_Get function. Do you think an extra function which does the entire file rewrite should be included (`Version_File_Update`) ?

stephanh commented:
Nah. The version files are quite specific to the toolset, i.e. version.sbt or `<name>.cabal`. The function as it is seems fine to me.

wrouesnel commented:
Ok so that's properly passing tests now!

stephanh commented:
:+1:
